listbox System.StackOverflowException: The requested operation caused a stack overflow

### Describe the bug

Listbox scroll with System.StackOverflowException: The requested operation caused a stack overflow:

mono-rt	[ERROR] FATAL UNHANDLED EXCEPTION: System.StackOverflowException: The requested operation caused a stack overflow.
mono-rt	at Avalonia.AvaloniaProperty.Equals (Avalonia.AvaloniaProperty) [0x00001] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.StyledElement.OnPropertyChanged (Avalonia.AvaloniaPropertyChangedEventArgs) [0x00012] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.Visual.OnPropertyChanged (Avalonia.AvaloniaPropertyChangedEventArgs) [0x00000] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.Layout.Layoutable.OnPropertyChanged (Avalonia.AvaloniaPropertyChangedEventArgs) [0x00000] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.Input.InputElement.OnPropertyChanged (Avalonia.AvaloniaPropertyChangedEventArgs) [0x00000] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.Controls.Control.OnPropertyChanged (Avalonia.AvaloniaPropertyChangedEventArgs) [0x00000] in <7d852e5c65c94d26afd58d22b2566069>:0
mono-rt	at Avalonia.Controls.ContentControl.OnPropertyChanged (Avalonia.AvaloniaPropertyChangedEventArgs) [0x00000] in <7d852e5c65c94d26afd58d22b2566069>:0
mono-rt	at Avalonia.Controls.ScrollViewer.OnPropertyChanged (Avalonia.AvaloniaPropertyChangedEventArgs) [0x00000] in <7d852e5c65c94d26afd58d22b2566069>:0
mono-rt	at Avalonia.Animation.Animatable.OnPropertyChangedCore (Avalonia.AvaloniaPropertyChangedEventArgs) [0x001f3] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.AvaloniaObject.RaisePropertyChanged<Avalonia.Vector> (Avalonia.AvaloniaProperty`1<Avalonia.Vector>,Avalonia.Data.Optional`1<Avalonia.Vector>,Avalonia.Data.BindingValue`1<Avalonia.Vector>,Avalonia.Data.BindingPriority,bool) [0x0000e] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.EffectiveValue`1<Avalonia.Vector>.NotifyValueChanged (Avalonia.PropertyStore.ValueStore,Avalonia.StyledProperty`1<Avalonia.Vector>,Avalonia.Vector) [0x0000e] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.EffectiveValue`1<Avalonia.Vector>.SetAndRaiseCore (Avalonia.PropertyStore.ValueStore,Avalonia.StyledProperty`1<Avalonia.Vector>,Avalonia.Vector,Avalonia.Data.BindingPriority,bool,bool) [0x000dc] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.ValueStore.SetCurrentValue<Avalonia.Vector> (Avalonia.StyledProperty`1<Avalonia.Vector>,Avalonia.Vector) [0x00014] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.AvaloniaObject.SetCurrentValue<Avalonia.Vector> (Avalonia.StyledProperty`1<Avalonia.Vector>,Avalonia.Vector) [0x00041] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.Controls.Presenters.ScrollContentPresenter.OnPropertyChanged (Avalonia.AvaloniaPropertyChangedEventArgs) [0x0002a] in <7d852e5c65c94d26afd58d22b2566069>:0
mono-rt	at Avalonia.Animation.Animatable.OnPropertyChangedCore (Avalonia.AvaloniaPropertyChangedEventArgs) [0x001f3] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.AvaloniaObject.RaisePropertyChanged<Avalonia.Vector> (Avalonia.AvaloniaProperty`1<Avalonia.Vector>,Avalonia.Data.Optional`1<Avalonia.Vector>,Avalonia.Data.BindingValue`1<Avalonia.Vector>,Avalonia.Data.BindingPriority,bool) [0x0000e] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.EffectiveValue`1<Avalonia.Vector>.NotifyValueChanged (Avalonia.PropertyStore.ValueStore,Avalonia.StyledProperty`1<Avalonia.Vector>,Avalonia.Vector) [0x0000e] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.EffectiveValue`1<Avalonia.Vector>.SetAndRaiseCore (Avalonia.PropertyStore.ValueStore,Avalonia.StyledProperty`1<Avalonia.Vector>,Avalonia.Vector,Avalonia.Data.BindingPriority,bool,bool) [0x000dc] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.EffectiveValue`1<Avalonia.Vector>.SetAndRaise (Avalonia.PropertyStore.ValueStore,Avalonia.PropertyStore.IValueEntry,Avalonia.Data.BindingPriority) [0x00008] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.ValueStore.ReevaluateEffectiveValue (Avalonia.AvaloniaProperty,Avalonia.PropertyStore.EffectiveValue,Avalonia.PropertyStore.IValueEntry,bool) [0x000b2] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.ValueStore.OnBindingValueChanged (Avalonia.PropertyStore.IValueEntry,Avalonia.Data.BindingPriority) [0x0001b] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.BindingEntryBase`2<Avalonia.Vector, Avalonia.Vector>.<SetValue>g__Execute|35_0 (Avalonia.PropertyStore.BindingEntryBase`2<Avalonia.Vector, Avalonia.Vector>,Avalonia.Data.BindingValue`1<Avalonia.Vector>) [0x000e6] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.BindingEntryBase`2<Avalonia.Vector, Avalonia.Vector>.SetValue (Avalonia.Data.BindingValue`1<Avalonia.Vector>) [0x00017] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.BindingEntryBase`2<Avalonia.Vector, Avalonia.Vector>.OnNext (Avalonia.Data.BindingValue`1<Avalonia.Vector>) [0x00000] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.Reactive.LightweightObservableBase`1<Avalonia.Data.BindingValue`1<Avalonia.Vector>>.PublishNext (Avalonia.Data.BindingValue`1<Avalonia.Vector>) [0x000db] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.Reactive.AvaloniaPropertyBindingObservable`2<Avalonia.Vector, Avalonia.Vector>.PublishValue (Avalonia.Vector) [0x00031] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.Reactive.AvaloniaPropertyBindingObservable`2<Avalonia.Vector, Avalonia.Vector>.PropertyChanged (object,Avalonia.AvaloniaPropertyChangedEventArgs) [0x0001d] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at (wrapper delegate-invoke) System.EventHandler`1<Avalonia.AvaloniaPropertyChangedEventArgs>.invoke_void_object_TEventArgs (object,Avalonia.AvaloniaPropertyChangedEventArgs) [0x00086] in <dae543f253a24af88fe499e3b9a40b9b>:0
mono-rt	at Avalonia.AvaloniaObject.RaisePropertyChanged<Avalonia.Vector> (Avalonia.AvaloniaProperty`1<Avalonia.Vector>,Avalonia.Data.Optional`1<Avalonia.Vector>,Avalonia.Data.BindingValue`1<Avalonia.Vector>,Avalonia.Data.BindingPriority,bool) [0x0002a] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.EffectiveValue`1<Avalonia.Vector>.NotifyValueChanged (Avalonia.PropertyStore.ValueStore,Avalonia.StyledProperty`1<Avalonia.Vector>,Avalonia.Vector) [0x0000e] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.EffectiveValue`1<Avalonia.Vector>.SetAndRaiseCore (Avalonia.PropertyStore.ValueStore,Avalonia.StyledProperty`1<Avalonia.Vector>,Avalonia.Vector,Avalonia.Data.BindingPriority,bool,bool) [0x000dc] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.ValueStore.SetCurrentValue<Avalonia.Vector> (Avalonia.StyledProperty`1<Avalonia.Vector>,Avalonia.Vector) [0x00014] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.AvaloniaObject.SetCurrentValue<Avalonia.Vector> (Avalonia.StyledProperty`1<Avalonia.Vector>,Avalonia.Vector) [0x00041] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.Controls.Presenters.ScrollContentPresenter.OnPropertyChanged (Avalonia.AvaloniaPropertyChangedEventArgs) [0x0002a] in <7d852e5c65c94d26afd58d22b2566069>:0
mono-rt	at Avalonia.Animation.Animatable.OnPropertyChangedCore (Avalonia.AvaloniaPropertyChangedEventArgs) [0x001f3] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.AvaloniaObject.RaisePropertyChanged<Avalonia.Vector> (Avalonia.AvaloniaProperty`1<Avalonia.Vector>,Avalonia.Data.Optional`1<Avalonia.Vector>,Avalonia.Data.BindingValue`1<Avalonia.Vector>,Avalonia.Data.BindingPriority,bool) [0x0000e] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.EffectiveValue`1<Avalonia.Vector>.NotifyValueChanged (Avalonia.PropertyStore.ValueStore,Avalonia.StyledProperty`1<Avalonia.Vector>,Avalonia.Vector) [0x0000e] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.EffectiveValue`1<Avalonia.Vector>.SetAndRaiseCore (Avalonia.PropertyStore.ValueStore,Avalonia.StyledProperty`1<Avalonia.Vector>,Avalonia.Vector,Avalonia.Data.BindingPriority,bool,bool) [0x000dc] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.EffectiveValue`1<Avalonia.Vector>.SetAndRaise (Avalonia.PropertyStore.ValueStore,Avalonia.PropertyStore.IValueEntry,Avalonia.Data.BindingPriority) [0x00008] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.ValueStore.ReevaluateEffectiveValue (Avalonia.AvaloniaProperty,Avalonia.PropertyStore.EffectiveValue,Avalonia.PropertyStore.IValueEntry,bool) [0x000b2] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.ValueStore.OnBindingValueChanged (Avalonia.PropertyStore.IValueEntry,Avalonia.Data.BindingPriority) [0x0001b] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.BindingEntryBase`2<Avalonia.Vector, Avalonia.Vector>.<SetValue>g__Execute|35_0 (Avalonia.PropertyStore.BindingEntryBase`2<Avalonia.Vector, Avalonia.Vector>,Avalonia.Data.BindingValue`1<Avalonia.Vector>) [0x000e6] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.BindingEntryBase`2<Avalonia.Vector, Avalonia.Vector>.SetValue (Avalonia.Data.BindingValue`1<Avalonia.Vector>) [0x00017] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.BindingEntryBase`2<Avalonia.Vector, Avalonia.Vector>.OnNext (Avalonia.Data.BindingValue`1<Avalonia.Vector>) [0x00000] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.Reactive.LightweightObservableBase`1<Avalonia.Data.BindingValue`1<Avalonia.Vector>>.PublishNext (Avalonia.Data.BindingValue`1<Avalonia.Vector>) [0x000db] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.Reactive.AvaloniaPropertyBindingObservable`2<Avalonia.Vector, Avalonia.Vector>.PublishValue (Avalonia.Vector) [0x00031] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.Reactive.AvaloniaPropertyBindingObservable`2<Avalonia.Vector, Avalonia.Vector>.PropertyChanged (object,Avalonia.AvaloniaPropertyChangedEventArgs) [0x0001d] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at (wrapper delegate-invoke) System.EventHandler`1<Avalonia.AvaloniaPropertyChangedEventArgs>.invoke_void_object_TEventArgs (object,Avalonia.AvaloniaPropertyChangedEventArgs) [0x00086] in <dae543f253a24af88fe499e3b9a40b9b>:0
mono-rt	at Avalonia.AvaloniaObject.RaisePropertyChanged<Avalonia.Vector> (Avalonia.AvaloniaProperty`1<Avalonia.Vector>,Avalonia.Data.Optional`1<Avalonia.Vector>,Avalonia.Data.BindingValue`1<Avalonia.Vector>,Avalonia.Data.BindingPriority,bool) [0x0002a] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.EffectiveValue`1<Avalonia.Vector>.NotifyValueChanged (Avalonia.PropertyStore.ValueStore,Avalonia.StyledProperty`1<Avalonia.Vector>,Avalonia.Vector) [0x0000e] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.EffectiveValue`1<Avalonia.Vector>.SetAndRaiseCore (Avalonia.PropertyStore.ValueStore,Avalonia.StyledProperty`1<Avalonia.Vector>,Avalonia.Vector,Avalonia.Data.BindingPriority,bool,bool) [0x000dc] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.PropertyStore.ValueStore.SetCurrentValue<Avalonia.Vector> (Avalonia.StyledProperty`1<Avalonia.Vector>,Avalonia.Vector) [0x00014] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.AvaloniaObject.SetCurrentValue<Avalonia.Vector> (Avalonia.StyledProperty`1<Avalonia.Vector>,Avalonia.Vector) [0x00041] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.Controls.Presenters.ScrollContentPresenter.OnPropertyChanged (Avalonia.AvaloniaPropertyChangedEventArgs) [0x0002a] in <7d852e5c65c94d26afd58d22b2566069>:0
mono-rt	at Avalonia.Animation.Animatable.OnPropertyChangedCore (Avalonia.AvaloniaPropertyChangedEventArgs) [0x001f3] in <9c60568d0b54415dad3590c449730446>:0
mono-rt	at Avalonia.AvaloniaObject.RaisePropertyChanged<Avalonia.Vector> (Avalonia.AvaloniaProperty`1<Avalonia.Vector>,Avalonia.Data.Optional`1<Avalonia.Vector>,Avalonia.Data.BindingValue`1<Avalonia.Vector>,Avalonia.Data.BindingPriority,bool) [0x0000e] in [<9c60568d0b54415dad3590c449730446>:0]


### To Reproduce

Listbox scroll

### Expected behavior

_No response_

### Avalonia version

11.3.2

### OS

_No response_

### Additional context

_No response_

--- a/src/Avalonia.Controls/ScrollViewer.cs
+++ b/src/Avalonia.Controls/ScrollViewer.cs
@@ -696,6 +696,14 @@ namespace Avalonia.Controls
 
         private static double Clamp(double value, double min, double max)
         {
+            // A NaN offset must never survive. Offset is two-way coerced between the ScrollViewer and its
+            // ScrollContentPresenter; because NaN != NaN, a NaN offset never compares equal, so the
+            // coerce/raise cycle never converges and recurses until it overflows the stack (see #21444).
+            if (double.IsNaN(value))
+            {
+                return min;
+            }
+
             return (value < min) ? min : (value > max) ? max : value;
         }
 

--- a/tests/Avalonia.Controls.UnitTests/ScrollViewerTests.cs
+++ b/tests/Avalonia.Controls.UnitTests/ScrollViewerTests.cs
@@ -45,6 +45,25 @@ namespace Avalonia.Controls.UnitTests
         }
 
         [Fact]
+        public void Setting_Offset_To_NaN_Does_Not_Cause_Infinite_Coerce_Recursion()
+        {
+            var target = new ScrollViewer
+            {
+                Template = new FuncControlTemplate<ScrollViewer>(CreateTemplate),
+                Content = "Foo",
+                Extent = new Size(100, 100),
+                Viewport = new Size(10, 10),
+            };
+
+            InitializeScrollViewer(target);
+
+            target.Offset = new Vector(0, double.NaN);
+
+            Assert.False(double.IsNaN(target.Offset.X));
+            Assert.False(double.IsNaN(target.Offset.Y));
+        }
+
+        [Fact]
         public void Test_ScrollToHome()
         {
             var target = new ScrollViewer

